### PR TITLE
WIP Suggestion creation macro

### PR DESCRIPTION
**Pull request by Jarcho**
This is a minimal and hacky implementation right now, but it should get the idea across.

A large number of suggestions in clippy are basically extracting a snippet of an expression (or multiple), adding something around it, and then replacing an expression. This is a proc-macro made to simplify that as much as possible. The syntax is a regular rust expression with meta-variables to insert snippets. It should handle keeping track of precedence and inserting parenthesis around snippets as needed, as well as keeping track of the final precedence 

Some examples:

```
expr_sugg!(&*$snip(e).foo);
expr_sugg!(&$mut(m) *foo);
expr_sugg!(&$mut(m) foo[$snip(e)]);
```

changelog: None

### CODE REVIEW

File: clippy_utils/src/lib.rs [hunk 1]
```diff
@@ -56,6 +56,81 @@ pub mod ty;
 pub mod usage;
 pub mod visitors;
 
+pub mod _internal {
```
Comment (llogiq) [inline clippy_utils/src/lib.rs:59]: Why the underscore? I would've thought the name makes it clear that this is not for public consumption. Perhaps add `#[doc(hidden)]` instead?

**Review by llogiq — COMMENTED** [review]
Cool idea! I like how you did this without using `syn`. I do wonder how much this actually simplifies things though. Perhaps add another commit that uses it in a few lints once the tests pass so we can see the wins to be had?

File: clippy_utils/src/lib.rs [hunk 2]
```diff
@@ -56,6 +56,81 @@ pub mod ty;
...
+            Some(x) if has_enclosing_paren(&x) => return x,
+            Some(x) => x,
+            None => return "(..)".into(),
+        };
+        match expr.kind {
+            ExprKind::Unary(..) | ExprKind::AddrOf(..) | ExprKind::Box(..) if position > Prefix => {
+                format!("({})", snip)
+            },
+            ExprKind::Closure(..) | ExprKind::Ret(..) | ExprKind::Break(..) | ExprKind::Yield(..)
+                if position > Closure =>
```
Comment (llogiq) [inline clippy_utils/src/lib.rs:105]: That would be equal to `position != Closure`, right?

Comment (Jarcho) [reply clippy_utils/src/lib.rs:105]: Yes.

File: clippy_utils/src/lib.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_utils/src/lib.rs:59]: It's temporary anyways. Everything used will be regular `pub` items once I'm done.

Comment (llogiq) [reply clippy_utils/src/lib.rs:59]: Ah, Ok then.

File: clippy_utils/src/lib.rs (hunk 2 above)
Comment (llogiq) [reply clippy_utils/src/lib.rs:106]: Then I'd prefer `!=`, because it's not susceptible to reordering.

Comment (Jarcho) [reply clippy_utils/src/lib.rs]: `ExprPostion` is already in precedence order and relies on it.

Comment (llogiq) [reply clippy_utils/src/lib.rs]: Ok then.

**Comment by llogiq** [comment]
Anything needed or should I review this now?

**Comment by camsteffen** [comment]
I think a good goal here would be for the macro to expand into code that could be written without the macro. This would take a lot of complexity out of the macro implementation.

For example, I would like to be able to write:

```rust
let sugg = Suggestion::borrow( // adds '&' and parenthesis according to arg precedence
    Suggestion::from_expr(cx, expr) // gets snippet
)
.position(cx, expr.hir_id) // adds outer parenthesis if needed when replacing `expr`
.to_string();
```

**Comment by Jarcho** [comment]
That would strictly be more work. Currently it's being handled like a format string with the addition of keeping track of what precedence level snippets are being inserted at, and what the lowest precedence level in the expression is. 

> ```rust
> let sugg = Suggestion::borrow( // adds '&' and parenthesis according to arg precedence
>     Suggestion::from_expr(cx, expr) // gets snippet
> )
> .position(cx, expr.hir_id) // adds outer parenthesis if needed when replacing `expr`
> .to_string();
> ```

This kind of output requires building an ast as something like `x * $snip * y` and `x + $snip * y` would have to build the suggestion in different orders (the first being `mul(mul(x, $snip), y)` the second being `add(x, mul($snip, y))`).

That being said keeping `Sugg` around for more complicated things is still useful. (e.g. conditionally negating a suggestion)

**Comment by Jarcho** [comment]
Just playing around with syntax options:

1. `$expr(expr1) + $expr(expr2)`
2. `$expr + $expr, expr1, expr2`
3. `{expr} + {expr}, expr1, expr2`

There's also another option where the type of the expansion has a default value, and only needs to be specified sometimes:

1. `$(expr1) + &$mut(mut1) foo`
2. `$_ + &$mut foo, expr1, mut1`
3. `{} + &{mut} foo, expr1, mut1`

I like the look of the format string versions, but it does mean any block expressions would need double curly braces. (e.g. `|x| { x }` would need to be `|x| {{ x }}`)

The types could also be specified after out of band (e.g. `{} = &{} foo, expr1, mut: mut1`), but that would require a second pass over the input as the trailing arguments would have to be parsed first.

**Comment by llogiq** [comment]
I still wonder whether the macro is going to pull its weight in terms of complexity. While we have a number of rather complex suggestions, I doubt they add up to 1.2k lines of code.

**Comment by Jarcho** [comment]
I have a few plans to shrink the amount of code, so there's still hope there.

Ultimately I'd like an interface along the lines of

```rust
lint_expr(cx, LINT_NAME, msg, expr, Applicability::MachineApplicable, sugg!($expr(foo) + 1));
```

This would cover a good chunk of lints which just replace one expression with another. Cover everything else with `span_lint_and_then` with some helper function to take the result of the `sugg!` macro.

**Comment by llogiq** [comment]
Sorry for letting you wait, I had a very busy week. This should be good to merge after yet another rebase.

**Comment by llogiq** [comment]
Sorry, I had a busy week. But another question arised: How does this mesh with the new [diagnostics translation effort](https://blog.rust-lang.org/inside-rust/2022/08/16/diagnostic-effort.html) (which may include clippy in the future)?

**Comment by Jarcho** [comment]
As far as I can tell this shouldn't be any more work than what we would currently need to do.

**Comment by xFrednet** [comment]
Since there hasn't been any activity since the last comment asking about the status, I'm going to close this PR. @Jarcho you're welcome to reopen it or create a new one, if you're interested in continuing this work :D

**Pull request closed without merging**